Per-CPU policy stats

jrajahalme opened the pull request:
Move policy stats to a global per-CPU LRU map. This eliminates memory trashing and associated contention on the bpf datapath packet forwarding path.

Map size defaults to `65536`, older entries will be cleared out when more space is needed for new ones. This size is configurable via the new command line option `--bpf-policy-stats-map-max` between the minimum of `1024` and maximum of `16777216`. Equivalent Helm option is `bpf.policyStatsMapMax`. The actual map size is rounded down to the closest even multiple of "possible CPUs" in the system, and is observable with the `cilium-dbg status --verbose` command.

In order to retain existing behavior useful in policy bug triaging we clear the stats for any policy entry when the entry is updated. This way the policy stats after a policy update are still reported as starting from zeroes.

Policy stats are still reported via the same `cilium-dbg bpf policy get` API as before, so the only visible change is that if there is traffic hitting thousands of different policy rules, the older ones will be cleared as space is needed for new ones, whereas previously policy stats were cleared only when the endpoint's policy map entry was updated.

Unavailable stats are printed as `-` by `cilium-dbg bpf policy get`. Stats can be unavailable if evicted by LRU or not initialized to zero (when debug is disabled).

Address an existing TODO in `bpf/lib/policy.h`:
```
#ifdef POLICY_ACCOUNTING
	/* FIXME: Use per cpu counters */
```

jrajahalme committed:
policymap: Simplify prefix length handling

Replace map lookup with a bit shift and use uint8 type for prefix lengths
where possible.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
option: Fix test

Policy map size over max should not error out, but to clamp the value to
the max.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
daemon: Fix policy map text in daemon status

Policy maps are per endpoint, not global. Fix the label text for policy
map (max) size to "Endpoint policy" from "Global policy".

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joestringer commented on cilium-dbg/cmd/bpf_policy_get.go:
```diff
@@ -177,7 +177,7 @@ func formatMap(w io.Writer, statsMap []policymap.PolicyEntryDump) {
 	for _, stat := range statsMap {
-		prefixLen := stat.Key.Prefixlen - policymap.StaticPrefixBits
+		prefixLen := stat.Key.GetPrefixLen()
 		id := identity.NumericIdentity(stat.Key.Identity)
```
How does this command continue to print stats if the underlying fields are dropped from the entries that this code is looking up? I don't see other extra code here to look up the stats from the new map.

joestringer replied:
Ah, it looks like the code is changed in `pkg/maps/policymap` as part of the dump command to issue an extra syscall for each policymap lookup to look up the corresponding policy key.

I would be interesting to understand how this behaves when exceeding the LRU. For instance it would be nice to indicate that stats are known-inaccurate (ie we never got information about packets passing through).

jrajahalme replied:
You are right, the policymap gets amended by the stats. My first version had the stats lookup only here in the `cilium-dbg`, but then it became simpler to amend the dump with the stats in `pkg/maps/policymap` instead.

With LRU map it is not really possible to tell the never had stats/no longer have stats cases apart. We already clear the stats whenever the policy map entries are updated, so we do not have 100% certainty of "keeping all the stats". Given large enough LRU map I'd think we are well covered, especially since the only way to get these stats is `cilium-dbg` on the node or a sysdump. If the sysdump is taken close to any incident (e.g., CI flake) then IMO we have very high likelihood to have the relevant stats.

Maybe one relevant thing to keep in mind that the stats are per policymap key (+ endpoint), so if an endpoint has e.g., 3 policy map entries all the stats for that endpoint will be in 3x<nCPU> entries, even if there are thousands of connections involved.

joestringer replied:
Yeah I guess what I'm thinking is that it will be tempting to see "count=0" and assume this literally means no traffic hit that policy entry, then draw wrong inference from that observation. On the other hand it seems like it would be more generally accurate to report "count=unknown".

(Extra points if we could get signal about lost stats, but I don't think LRU will give us this today either :/ )

jrajahalme replied:
Unavailable stats are now printed as `-` by `cilium-dbg bpf policy get`. Stats can be unavailable if evicted by LRU or not initialized to zero (when debug is disabled).

Example output:
```
Endpoint ID: 3027
Path: /sys/fs/bpf/tc/globals/cilium_policy_v2_03027

POLICY   DIRECTION   IDENTITY   PORT/PROTO   PROXY PORT   AUTH TYPE   BYTES   PACKETS   PREFIX
Allow    Ingress     0          ANY          NONE         disabled    0       0         0
Allow    Ingress     1          ANY          NONE         disabled    -       -         0
Allow    Egress      0          ANY          NONE         disabled    3092    32        0
```

giorio94 reviewed (approved):
/lgtm for my codeowners.

asauber commented on pkg/option/config.go:
```diff
@@ -3467,6 +3475,16 @@ func (c *DaemonConfig) checkMapSizeLimits() error {
 	}
 
+	if c.PolicyStatsMapEntries < LimitTableMin {
+		return fmt.Errorf("specified PolicyStatsMap max entries %d must exceed minimum %d",
```
This should say "must be greater or equal to"

jrajahalme replied:
Fixed this by adding a new commit that does the same change (x7) for all instances of this.

asauber reviewed (dismissed):
Looks good for config, besides one error message nit

YutaroHayakawa commented on pkg/maps/policymap/policymap.go:
```diff
@@ -347,9 +354,15 @@ func newDenyEntry(key PolicyKey) PolicyEntry {
 	entry := newAllowEntry(key, proxyPortPriority, authReq, proxyPort)
+	if option.Config.Debug {
+		pm.stats.ClearStat(pm.epID, key)
+	}
```
Is this the case for the backwards compatibility mentioned in the commit message? Why do we only clear the stats in debug mode?

jrajahalme replied:
Clearing of stats is for backwards compatibility, as previously stats for a specific policy entry were cleared as a side effect of updating that entry. Explicit clearing is a separate system call, so this adds cost to the management of the policy maps. Generally we only care about the policy stats if we want to triage a problem, and in that case we tend to want to have dubug enabled to get sufficient visibility to what has happened. So we only clear the stats in the debug mode to only carry the additional cost of explicitly clearing the stats when we are likely to need the stats to have been cleared with the previous policy map entry update.

YutaroHayakawa commented on pkg/maps/policymap/policymap.go:
```diff
@@ -362,9 +375,15 @@ func (pm *PolicyMap) Allow(trafficDirection trafficdirection.TrafficDirection, i
 	entry := newDenyEntry(key)
+	if option.Config.Debug {
+		pm.stats.ClearStat(pm.epID, key)
+	}
```
Same here.

YutaroHayakawa commented on pkg/maps/policymap/policymap.go:
```diff
@@ -434,26 +453,55 @@ func (pm *PolicyMap) DumpToSlice() (PolicyEntriesDump, error) {
+	}
+	statsMap, err := OpenStatsMap()
+	if err != nil {
+		log.WithError(err).Error("Failed to open StatsMap")
```
Is it ok to just fallthrough on these error cases and set `epID` to zero?

jrajahalme replied:
Yes, as the stats map lookup with the zero epID will produce zero results.

YutaroHayakawa reviewed (approved):
Reviewed BPF and map userspace. Looks good to me, but I'd like Joe or other policy folks to double check since this area is out of my expertise. A few questions from me.

jrajahalme committed:
option: Fix error message language

Be more precise in the error messages complaining about a given map size
being too small. Instead of stating that the configured number of entries
"must exceed minimum" number of entries, state that the configured number
of enties "must be greater or equal to" the minimum number of entries.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

gandro commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -443,6 +443,11 @@ data:
+{{- if hasKey .Values.bpf "policyStatsMapMax" }}
+  # bpf-policy-stats-map-max specifies the maximum number of entries in global
+  # policy stats map
+  bpf-policy-stats-map-max: "{{ .Values.bpf.policyStatsMapMax }}"
```
```suggestion
  bpf-policy-stats-map-max: "{{ .Values.bpf.policyStatsMapMax | int }}"
```

See https://github.com/cilium/cilium/issues/34010 for details on why

jrajahalme replied:
Fixed, thanks!

gandro reviewed (approved):
ACK for Helm/sig-k8s

bimmlerd commented on pkg/maps/policymap/statsmap.go:
```diff
@@ -0,0 +1,148 @@
+
+// PolicyStatsMap maps endpoint IDs to the fd for the program which
+// implements its policy.
+type StatsMap struct {
```
If I understand correctly, this is a new map - for new code, I'd expect a map cell (like the examples in `pkg/maps/cells.go`). Is there a reason this isn't a cell?

jrajahalme replied:
No real reason, added a new commit with a Cell that handles the new map and the config of the Endpoint's policymap.

jrajahalme replied:
OK, the real reason was that it was a ton of work. Done.

bimmlerd replied:
Thanks for doing it anyway.

bimmlerd commented on pkg/option/config.go:
```diff
@@ -1493,6 +1496,11 @@ type DaemonConfig struct {
+	// PolicyStatsMapEntries is the maximum number of stats entries kept
+	// globally in the node in an LRU map, where oldest entries are recycled
+	// to make space for new ones.
+	PolicyStatsMapEntries int
```
Let's also avoid adding new stuff to `option.Config` if we can, please. If you make the policy stat map a cell, I think you can declare your config there and won't have to add a new field here.

I haven't checked whether you need access to this field somewhere which doesn't have access to the hive - if so, that may be trickier.

jrajahalme replied:
Moved to the new Cell.

bimmlerd commented on daemon/cmd/daemon.go:
```diff
@@ -315,7 +316,13 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
 	ctmap.InitMapInfo(option.Config.EnableIPv4, option.Config.EnableIPv6, option.Config.EnableNodePort)
-	policymap.InitMapInfo(option.Config.PolicyMapEntries)
+
+	nCPU, err := ebpf.PossibleCPU()
```
If you make the policy stat map a cell, you could put this into its start hook and avoid further polluting `daemon.go`.

jrajahalme replied:
Done.

bimmlerd reviewed (changes requested):
I haven't reviewed much - but in my opinion new code should use the hive/cell infra.

jrajahalme committed:
policymap: Add policymap.Factory

Add policymap.Factory that opens policy stats map once and uses the same
stats map with all endpoint policy maps. Endpoint policy maps are opened
or created via the Factory methods.

Daemon passes the shared policymap.Factory to all endpoints. In tests it
is passed as nil as we already have mechanisms for not actually creating
endpoint policy maps in tests. This can later be modified to pass in a
mock policy map factory in tests.

Consolidate endpoint policy map operations in to the new Factory.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
bpf: Move policy stats to a PERCPU map

Move policy stats to a new per-CPU LRU hash map. This allows for policy
stats updates without contention between CPUs.

Use of a global LRU map saves memory with the consequence that least
recently used policy stats are cleared.

Policy stats entry is zeroed on corresponding policy map entry update
when debug is enabled to keep the existing behavior where newly inserted
policies have zeroed packet and byte counts.

Unavailable stats are printed as '-' in 'cilium-dbg bpf policy
get'. Stats can be unavailable if evicted by LRU or not initialized to
zero (when debug is disabled).

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

bimmlerd reviewed (approved):
🚀 (approving for `cilium/fqdn` and `cilium/contributing` only!)

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -782,9 +782,6 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.Int(option.NeighMapEntriesGlobalName, option.NATMapEntriesGlobalDefault, "Maximum number of entries for the global BPF neighbor table")
 	option.BindEnv(vp, option.NeighMapEntriesGlobalName)
 
-	flags.Int(option.PolicyMapEntriesName, policymap.MaxEntries, "Maximum number of entries in endpoint policy map (per endpoint)")
```
We should document this flag removal in the upgrade guide.

jrajahalme replied:
This flag is not removed, but merely moved to the new Cell. There should be no upgrade impact because of this :-)

aanm commented on pkg/maps/policymap/statsmap.go:
```diff
@@ -0,0 +1,173 @@
+	}
+
+	if !errors.Is(err, unix.ENOENT) {
+		m.log.Warn("Error getting policy stats", "error", err, logfields.BPFMapKey, statsKey)
```
s/`error`/`logfields.Error`

jrajahalme replied:
Done.

aanm commented on pkg/maps/policymap/statsmap.go:
```diff
@@ -0,0 +1,173 @@
+		return nil
+	}
+
+	m.log.Warn("Error deleting policy stats", "error", err, logfields.BPFMapKey, statsKey)
```
s/`error`/`logfields.Error`

jrajahalme replied:
Done.

aanm commented on pkg/maps/policymap/statsmap.go:
```diff
@@ -0,0 +1,173 @@
+	err := m.Update(&statsKey, zeroValue, 0)
+
+	if err != nil {
+		m.log.Warn("Error zeroing policy stats", "error", err, logfields.BPFMapKey, statsKey)
```
s/`error`/`logfields.Error`

jrajahalme replied:
Done.
